### PR TITLE
RegressionModel and sthocastic gradient

### PR DESCRIPTION
**Pull request by RogerioOMDS**
Pouca alteração do código, apenas a separação do arquivo dos modelos de regressão do arquivo das funções e métodos.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "julia.environmentPath": "c:\\Users\\Acer\\Desktop\\roger\\TCC\\MultObjNLPModels.jl"
+}
```
Comment (abelsiqueira) [inline .vscode/settings.json]: Esse arquivo não deve ser adicionado.

File: example.jl [hunk 2]
```diff
@@ -54,4 +54,4 @@ function NLPModels.hprod!(nlp :: ExampleMONLP, i :: Integer, x :: AbstractVector
     Hv[2] = -400 * x[1] * v[1] + 200 * v[2]
   end
   return Hv
-end
\ No newline at end of file
+end
```
Comment (abelsiqueira) [inline example.jl:57]: Não adicione essa modificação.

File: src/MultObjNLPModels.jl [hunk 3]
```diff
@@ -6,4 +6,6 @@ include("meta.jl")
 include("counters.jl")
 include("api.jl")
 
+include("gradestoc.jl")
+include("regressaomodel.jl")
```
Comment (abelsiqueira) [inline src/MultObjNLPModels.jl]: Mude os nomes de tudo para inglês. `regression_model` e `sthocastic_gradient` para os arquivos.
`RegressionModel` para a struct.

File: src/gradestoc.jl [hunk 4]
```diff
@@ -0,0 +1,131 @@
+export grad_desc, grad_estoc, MultObjNLPModels.obj, MultObjNLPModels.grad!
+
+
+function grad_desc(nlp::RegressaoModel;
```
Comment (abelsiqueira) [inline src/gradestoc.jl]: Esse método não deveria estar aqui. Ele não é importante para o problema, então coloque no seu exemplo.

File: src/gradestoc.jl [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    max_eval=100000,
+    max_time = 60.0,
+    atol = 1e-10,
+    rtol = 1e-10 
+    )
+    
+    start_time = time()
+    iter = 0
+    β = nlp.meta.x0
+    ∇ℓ(x,y,β) = ForwardDiff.gradient(β->nlp.ℓ(y, nlp.h(x,β)), β) #rever
```
Comment (abelsiqueira) [inline src/gradestoc.jl]: É pra usar o `grad` do modelo.

File: src/gradestoc.jl [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+  end
+  
+
+function grad_estoc(nlp::RegressaoModel;
+                      γ = 1e-2,
+                      max_eval=100000,
+                      max_time = 60.0,
+                      atol = 1e-10,
+                      rtol = 1e-10 
+                      )
```
Comment (abelsiqueira) [inline src/gradestoc.jl]:
```suggestion
function sthocastic_gradient(
    nlp::RegressionModel;
    γ = 1e-2,
    max_eval=10000,
    max_time = 60.0,
    atol = 1e-8,
    rtol = 1e-8 
)
```

File: src/gradestoc.jl [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+  @info log_header([:iter, :f],
+                   [Int, Float64],
+                   hdr_override=Dict(:f => "f(β)"))
+    
+  @info log_row(Any[iter, f])
+    
+  # while !(solved || tired)
+  for k = 1:10n
+    for i in shuffle(1:n)
+      β -= γ * MultObjNLPModels.grad!(nlp, i, β)
```
Comment (abelsiqueira) [inline src/gradestoc.jl]: `grad!`  não está seguindo o formato correto.

File: src/gradestoc.jl [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+  
+    # return GenericExecutionStats(status, nlp; 
+    #                               solution = β, 
+    #                               objective=f, 
+    #                               elapsed_time=Δt,
+    #                               iter=iter)
+
+    y_pred = [nlp.h(nlp.X[i,:],β) > 0.5 ? 1.0 : 0.0 for i=1:n ]
+    return y_pred
+  end
```
Comment (abelsiqueira) [inline src/gradestoc.jl]: Já tínhamos corrigido isso.
A indentação está incorreta também.

File: src/gradestoc.jl [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+  
+  
+  
+  function MultObjNLPModels.obj(nlp :: RegressaoModel, i :: Integer, β :: AbstractVector)
+    # @lencheck alguma coisa
+    return nlp.ℓ(nlp.y[i], h(nlp.X[i,:],β )) # ℓ( yᵢ, h(xᵢᵀβ) ) 
+  end
+  
+  function MultObjNLPModels.grad!(nlp :: RegressaoModel, i :: Integer, x :: AbstractVector #= , g :: AbstractVector =#)
```
Comment (abelsiqueira) [inline src/gradestoc.jl]: Vide autodiff_model.jl no NLPModels.jl

File: src/regressaomodel.jl [hunk 10]
```diff
@@ -0,0 +1,33 @@
+export RegressaoModel, RegressaoLinearModel, RegressaoLogisticaModel
+
+mutable struct RegressaoModel <: AbstractMultObjModel
```
Comment (abelsiqueira) [inline src/regressaomodel.jl]: Corrigir tudo pra inglês e corrigir indentação.

File: src/regressaomodel.jl [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    n, p = size(X)
+    
+    meta = MultObjNLPMeta(p, n)
+    counters = MultObjCounters(n)
+    return RegressaoModel(meta, counters, X, y, h, ℓ)
+  end
+  
+  function RegressaoLinearModel(X, y)
+    h(x, β) = x*β
+    ℓ(y, ŷ) = sum((y - ŷ).^2) / 2
```
Comment (abelsiqueira) [inline src/regressaomodel.jl]:
```suggestion
    ℓ(y, ŷ) = (y - ŷ)^2 / 2
```

File: src/regressaomodel.jl [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+function RegressaoModel(X, y, h, ℓ)
+    n, p = size(X)
+    
+    meta = MultObjNLPMeta(p, n)
+    counters = MultObjCounters(n)
+    return RegressaoModel(meta, counters, X, y, h, ℓ)
+  end
+  
+  function RegressaoLinearModel(X, y)
+    h(x, β) = x*β
```
Comment (abelsiqueira) [inline src/regressaomodel.jl]:
```suggestion
    h(x, β) = dot(x, β)
```

File: src/regressaomodel.jl [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    ℓ(y, ŷ) = sum((y - ŷ).^2) / 2
+    return RegressaoModel(X, y, h, ℓ) 
+  end
+  
+  function RegressaoLogisticaModel(X, y)
+    h(x, β) = 1 / (1 + exp(-x'*β))
```
Comment (abelsiqueira) [inline src/regressaomodel.jl]:
```suggestion
    h(x, β) = 1 / (1 + exp(-dot(x, β)))
```

**Review by abelsiqueira — CHANGES_REQUESTED** [review]
Fiz algumas várias modificações. Você disse que estava com erro?

File: example.jl [hunk 14]
```diff
@@ -54,4 +54,5 @@ function NLPModels.hprod!(nlp :: ExampleMONLP, i :: Integer, x :: AbstractVector
...
+
```
Comment (abelsiqueira) [inline example.jl:58]:
```suggestion
end
```

File: src/MultObjNLPModels.jl [hunk 15]
```diff
@@ -6,4 +6,6 @@ include("meta.jl")
 include("counters.jl")
 include("api.jl")
 
-end # module
+include("regression_model.jl")
+include("sthocastic_gradient.jl")
+end # module
```
Comment (abelsiqueira) [inline src/MultObjNLPModels.jl]:
```suggestion

end # module
```

File: src/sthocastic_gradient.jl [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+export sthocastic_gradient
+
+function sthocastic_gradient(
+    nlp::AbstractMultObjModel;
+    learning_rate::Symbol = :optimal, 
+    γ::Float64 = 1e-2, 
+    α::Float64 = 1e-2, # for penalty
+    ρ::Float64 = 0.85, # just for elasticnet
+    penalty::Symbol = :l2, 
+    max_eval::Int=50,
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
    max_eval::Int = 50,
```

File: src/sthocastic_gradient.jl [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+    max_time::Float64 = 60.0,
+    atol::Float64 = 1e-8,
+    rtol::Float64 = 1e-8,
+    power_t::Float64 = 1e-2
+  )
+
+  iter = 0
+  start_time = time()
+  β = nlp.meta.x0
+  βavg = similar(β) # test later
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
  βavg = similar(β)
```

File: src/sthocastic_gradient.jl [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+  n = nlp.meta.nobj
+  g = similar(β)
+  f = NLPModels.obj(nlp, β)
+  NLPModels.grad!(nlp,β,g)
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
  NLPModels.grad!(nlp, β, g)
```

File: src/sthocastic_gradient.jl [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+  
+  Δt = time() - start_time
+
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
```

File: src/sthocastic_gradient.jl [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+  @info log_row(Any[iter, f])
+
+  
+  P = if penalty == :l2
+    β -> β 
+  elseif penalty == :l1
+    β -> sign.(β)
+  elseif penalty == :elasticnet
+    β -> ρ * β + (1-ρ) * sign.(β)
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
    β -> ρ * β + (1 - ρ) * sign.(β)
```

File: src/sthocastic_gradient.jl [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+  end
+  
+
+  for k = 1:max_eval
+  # while !(γ < 0.01 || tired) # sounds good. Doesnt work
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
```

File: src/sthocastic_gradient.jl [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+    elseif learning_rate == :constant
+      γ = γ    
+    end
+
+    βavg .= 0 # test later
+    for i in shuffle(1:n)
+      β -= γ * (α * P(β) + grad!(nlp, i, β, g))
+      βavg += β # test later
+    end
+    βavg = βavg/n # test later
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
    βavg .= 0
    for i in shuffle(1:n)
      β -= γ * (α * P(β) + grad!(nlp, i, β, g))
      βavg += β
    end
    βavg = βavg / n
```

File: src/sthocastic_gradient.jl [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+  end
+  
+
+  for k = 1:max_eval
+  # while !(γ < 0.01 || tired) # sounds good. Doesnt work
+
+    if learning_rate == :optimal
+      γ = 1/(α * (1e3 + iter)) 
+    elseif learning_rate == :invscaling 
+      γ = 1e-2 / (iter+1)^(power_t)
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
      γ = 1e-2 / (iter + 1)^power_t
```

File: src/sthocastic_gradient.jl [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+      β -= γ * (α * P(β) + grad!(nlp, i, β, g))
+      βavg += β # test later
+    end
+    βavg = βavg/n # test later
+
+    Δt = time() - start_time
+    tired = (Δt > max_time) 
+    solved = (γ < 1e-2)
+    itermax = iter > max_eval
+    iter+=1
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
    iter += 1
    itermax = iter > max_eval
```

File: src/sthocastic_gradient.jl [hunk 25]
```diff
@@ -0,0 +1,95 @@
...
+  elseif penalty == :elasticnet
+    β -> ρ * β + (1-ρ) * sign.(β)
+  end
+  
+
+  for k = 1:max_eval
+  # while !(γ < 0.01 || tired) # sounds good. Doesnt work
+
+    if learning_rate == :optimal
+      γ = 1/(α * (1e3 + iter))
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
      γ = 1 / (α * (1e3 + iter)) 
```

File: src/sthocastic_gradient.jl [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+  for k = 1:max_eval
+  # while !(γ < 0.01 || tired) # sounds good. Doesnt work
+
+    if learning_rate == :optimal
+      γ = 1/(α * (1e3 + iter)) 
+    elseif learning_rate == :invscaling 
+      γ = 1e-2 / (iter+1)^(power_t)
+    elseif learning_rate == :constant
+      γ = γ    
+    end
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
    end
```

File: src/sthocastic_gradient.jl [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+    βavg .= 0 # test later
+    for i in shuffle(1:n)
+      β -= γ * (α * P(β) + grad!(nlp, i, β, g))
+      βavg += β # test later
+    end
+    βavg = βavg/n # test later
+
+    Δt = time() - start_time
+    tired = (Δt > max_time) 
+    solved = (γ < 1e-2)
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
    tired = Δt > max_time
    solved = γ < 1e-2
```

File: src/sthocastic_gradient.jl [hunk 28]
```diff
@@ -0,0 +1,95 @@
...
+  else
+    :unknown
+  end
+
+  return GenericExecutionStats(status, nlp;  
+                                solution = β,
+                                solver_specific=Dict(:βavg=>βavg),
+                                elapsed_time=Δt,
+                                iter=iter
+                                )
```
Comment (abelsiqueira) [inline src/sthocastic_gradient.jl]:
```suggestion
  return GenericExecutionStats(status, nlp;  
                               solution=β,
                               solver_specific=Dict(:βavg => βavg),
                               elapsed_time=Δt,
                               iter=iter
                               )
```

**Review by abelsiqueira — CHANGES_REQUESTED** [review]
Thanks for the update. I made some modifications that should be easy to accept (Add suggestion to batch and merge everything afterwards), but please take a look at each change.
Additional changes:
- Use 2 space indentation.
- Don't leave trailing whitespaces (there are packages to warn you).
- Remove exemplosrogerio.jl, it's not supposed to be in the repo.
- `max_eval` and `max_iter` are not correct. You should have
  - `max_iter` for the maximum of iteration.
  - `max_eval` for the maximum of function evaluations.
  - Here, remove `max_eval`, use only `max_iter`.

There are a couple more things, but I'll wait for the update first.
